### PR TITLE
feat(crypto-ffi-bindings): Save/Load dehydrated pickle key

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Draft because depends on https://github.com/matrix-org/matrix-rust-sdk/pull/4383

- [ ] Public API changes documented in changelogs (optional)

Signed-off-by:

### CODE REVIEW

File: bindings/matrix-sdk-crypto-ffi/src/dehydrated_devices.rs [hunk 1]
```diff
@@ -73,20 +78,18 @@ impl DehydratedDevices {
...
             inner: ManuallyDrop::new(self.runtime.block_on(self.inner.rehydrate(
-                &key,
+                key,
                 &device_id,
                 device_data,
             ))?),
         }
         .into();
 
-        key.zeroize();
```
Comment (BillCarsonFr) [inline bindings/matrix-sdk-crypto-ffi/src/dehydrated_devices.rs:88]: The `DehydratedDeviceKey` derives `ZeroizeOnDrop` so I believe this is not needed anymore.
I have a question regarding the `pickle_key` parameter, should not it be zeroized too?

Comment (poljar) [reply bindings/matrix-sdk-crypto-ffi/src/dehydrated_devices.rs:88]: Both observations are correct.

File: bindings/matrix-sdk-crypto-ffi/src/dehydrated_devices.rs [hunk 2]
```diff
@@ -25,6 +26,8 @@ pub enum DehydrationError {
     Store(#[from] matrix_sdk_crypto::CryptoStoreError),
     #[error("The pickle key has an invalid length, expected 32 bytes, got {0}")]
     PickleKeyLength(usize),
+    #[error(transparent)]
+    Rand(#[from] rand::Error),
```
Comment (BillCarsonFr) [inline bindings/matrix-sdk-crypto-ffi/src/dehydrated_devices.rs:30]: Needed for the ffi `DehydratedDeviceKey::new()`

File: bindings/matrix-sdk-crypto-ffi/src/dehydrated_devices.rs [hunk 3]
```diff
@@ -180,15 +215,20 @@ impl From<dehydrated_device::put_dehydrated_device::unstable::Request>
...
 }
 
-fn get_pickle_key(pickle_key: &[u8]) -> Result<DehydratedDeviceKey, DehydrationError> {
-    let pickle_key_length = pickle_key.len();
+#[cfg(test)]
+mod tests {
+    use crate::DehydratedDeviceKey;
+
+    #[test]
+    fn test_dehydrated_key() {
```
Comment (andybalaam) [inline bindings/matrix-sdk-crypto-ffi/src/dehydrated_devices.rs:223]: How about `test_creating_dehydrated_key_from_slice` ?

File: bindings/matrix-sdk-crypto-ffi/src/lib.rs [hunk 4]
```diff
@@ -822,6 +827,39 @@ impl TryFrom<matrix_sdk_crypto::store::BackupKeys> for BackupKeys {
...
+impl DehydratedDeviceKey {
+    /// Generates a new random pickle key.
+    pub fn new() -> Result<Self, DehydrationError> {
+        let inner = InnerDehydratedDeviceKey::new()?;
+        Ok(inner.into())
+    }
+
+    /// Creates a new dehydration pickle key from the given slice.
+    ///
+    /// Fail if the slice length is not 32.
```
Comment (andybalaam) [inline bindings/matrix-sdk-crypto-ffi/src/lib.rs:845]: Can we have a test for the failure case?

File: bindings/matrix-sdk-crypto-ffi/src/dehydrated_devices.rs (hunk 3 above)
Comment (andybalaam) [reply bindings/matrix-sdk-crypto-ffi/src/dehydrated_devices.rs:223]: Or `test_roundtrip_dehydrated_device_key` maybe

File: crates/matrix-sdk-crypto/src/store/mod.rs [hunk 5]
```diff
@@ -775,6 +778,19 @@ impl DehydratedDeviceKey {
         Ok(Self { inner: key })
     }
 
+    /// Creates a new dehydration pickle key from the given slice.
+    ///
+    /// Fail if the slice length is not 32.
+    pub fn from_slice(slice: &[u8]) -> Result<Self, DehydrationError> {
```
Comment (andybalaam) [inline crates/matrix-sdk-crypto/src/store/mod.rs:784]: It would be good to have a test for this too.

**Review by andybalaam — APPROVED** [review]
Looks good, a couple of suggestions.

**Pull request merged**